### PR TITLE
Add kinds for binary operators and fix FPs on S1764

### LINKED ISSUE #149: FP on S1764
`<<` operator should be excluded when used with literals

https://peach.sonarsource.com/project/issues?id=sonargo%3Amoby&issues=AWIygRAorCUbrBGdP4lI&open=AWIygRAorCUbrBGdP4lI

### PR DESCRIPTION
**Pull request by saberduck**
Fix #149

### CODE REVIEW

File: uast-generator-java/src/main/java/org/sonar/uast/generator/java/Generator.java [hunk 1]
```diff
@@ -255,33 +255,14 @@ private static UastNode newUastNode(Tree tree, List<UastNode> children) {
...
-        result.add(UastNode.Kind.OPERATOR_LOGICAL_AND);
-        break;
-      case "||":
-        result.add(UastNode.Kind.OPERATOR);
-        result.add(UastNode.Kind.OPERATOR_LOGICAL_OR);
-        break;
+        result.add(kind);
+      }
+    }
+    switch (tokenText) {
```
Comment (SonarTech) [inline uast-generator-java/src/main/java/org/sonar/uast/generator/java/Generator.java]: Code Smell Code Smell: Replace this "switch" statement by "if" statements to increase readability. ([squid:S1301](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1301))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=188&issues=AWJtpw4XsgPCf6tZNzw_&open=AWJtpw4XsgPCf6tZNzw_)

[ISSUE_KEY]: AWJtpw4XsgPCf6tZNzw_

File: common-rule-engine/src/main/java/org/sonar/uast/UastNode.java [hunk 2]
```diff
@@ -139,19 +139,30 @@ public String toString() {
...
+    OPERATOR_LEFT_SHIFT,
+    OPERATOR_RIGHT_SHIFT,
     OPERATOR_EQUAL,
     OPERATOR_LOGICAL_AND,
     OPERATOR_LOGICAL_OR,
-    OPERATOR_MULTIPLY,
     OPERATOR_NOT_EQUAL,
+    OPERATOR_LESS_THEN,
+    OPERATOR_LESS_OR_EQUAL,
+    OPERATOR_GREATER_THEN,
```
Comment (alban-auzeill) [inline common-rule-engine/src/main/java/org/sonar/uast/UastNode.java]: OPERATOR_GREATER_THAN

File: common-rule-engine/src/main/java/org/sonar/uast/UastNode.java [hunk 3]
```diff
@@ -139,19 +139,30 @@ public String toString() {
...
+    OPERATOR_BINARY_OR,
+    OPERATOR_BINARY_XOR,
+    OPERATOR_LEFT_SHIFT,
+    OPERATOR_RIGHT_SHIFT,
     OPERATOR_EQUAL,
     OPERATOR_LOGICAL_AND,
     OPERATOR_LOGICAL_OR,
-    OPERATOR_MULTIPLY,
     OPERATOR_NOT_EQUAL,
+    OPERATOR_LESS_THEN,
```
Comment (alban-auzeill) [inline common-rule-engine/src/main/java/org/sonar/uast/UastNode.java]: OPERATOR_LESS_THAN

File: common-rule-engine/src/test/files/checks/java/BinaryOperatorIdenticalExpressionsCheck/BinaryOperatorIdenticalExpressionsCheck.java [hunk 4]
```diff
@@ -23,7 +23,7 @@ void foo() {
     if(d != d) {} // Noncompliant
     int j,l;
     // SonarJava:exclude this case for bit masks
-    int k = 1 << 1; // Noncompliant
+    int k = 1 << 1; // XXXcompliant
```
Comment (alban-auzeill) [inline common-rule-engine/src/test/files/checks/java/BinaryOperatorIdenticalExpressionsCheck/BinaryOperatorIdenticalExpressionsCheck.java:26]: XXXcompliant => Compliant

File: uast-generator-java/src/main/java/org/sonar/uast/generator/java/UastNode.java [hunk 5]
```diff
@@ -69,28 +68,51 @@ public UastNode(Set<Kind> kinds, String nativeNode, @Nullable Token token, List<
...
     RETURN,
     STATEMENT,
     SWITCH,
     THEN,
     THROW,
     TYPE,
     ;
+
+    @Nullable
+    public final String token;
```
Comment (ghost) [inline uast-generator-java/src/main/java/org/sonar/uast/generator/java/UastNode.java]: I like the idea to simplify matching, but I don't like the way it is used to be checked. Could you had a method in the enum to match from token? `public boolean isKindFromToken(String token)` ?

I could use the same approach for compound assignments

File: uast-generator-java/src/main/java/org/sonar/uast/generator/java/Generator.java [hunk 6]
```diff
@@ -255,33 +255,14 @@ private static UastNode newUastNode(Tree tree, List<UastNode> children) {
   }
 
   private static void addTokenKinds(SyntaxToken tree, Set<UastNode.Kind> result) {
-    SyntaxToken token = tree;
-    switch (token.text()) {
-      // operators
-      case "+":
+    String tokenText = tree.text();
+    for (UastNode.Kind kind : UastNode.Kind.values()) {
+      if (tokenText.equals(kind.token)) {
```
Comment (ghost) [inline uast-generator-java/src/main/java/org/sonar/uast/generator/java/Generator.java]: `kind.isKindFromToken(tokenText)` ?

File: common-rule-engine/src/test/files/checks/java/BinaryOperatorIdenticalExpressionsCheck/BinaryOperatorIdenticalExpressionsCheck.java (hunk 4 above)
Comment (saberduck) [reply common-rule-engine/src/test/files/checks/java/BinaryOperatorIdenticalExpressionsCheck/BinaryOperatorIdenticalExpressionsCheck.java:26]: I used XXX to mark a difference with SonarJava as indicated in the comment above

File: uast-generator-go/goparser.go [hunk 7]
```diff
@@ -39,65 +39,78 @@ func (k Kind) String() string {
...
+	BREAK                     Kind = "BREAK"
+	CONTINUE                  Kind = "CONTINUE"
+	OPERATOR                  Kind = "OPERATOR"
+	OPERATOR_ADD              Kind = "OPERATOR_ADD"
+	OPERATOR_SUBTRACT         Kind = "OPERATOR_SUBTRACT"
+	OPERATOR_MULTIPLY         Kind = "OPERATOR_MULTIPLY"
+	OPERATOR_DIVIDE           Kind = "OPERATOR_DIVIDE"
+	OPERATOR_MODULO           Kind = "OPERATOR_MODULO"
+	OPERATOR_BINARY_AND       Kind = "OPERATOR_BINARY_AND"
+	OPERATOR_BINARY_AND_NOT   Kind = "OPERATOR_BINARY_AND_NOT"
```
Comment (ghost) [inline uast-generator-go/goparser.go:98]: I would add this one to common.UastNode.Kind, as being a super set of both languages, java and go

File: uast-generator-go/goparser.go [hunk 8]
```diff
@@ -236,18 +249,65 @@ func (t *UastMapper) computeBasicLitKinds(tok token.Token) []Kind {
...
+		return []Kind{OPERATOR, OPERATOR_LOGICAL_AND}
+		// ||
+	case token.LOR:
+		return []Kind{OPERATOR, OPERATOR_LOGICAL_OR}
+		// ==
 	case token.EQL:
 		return []Kind{OPERATOR, OPERATOR_EQUAL}
+		// <
+	case token.LSS:
+		return []Kind{OPERATOR, OPERATOR_LESS_THEN}
```
Comment (ghost) [inline uast-generator-go/goparser.go]: `THEN` -> `THAN`

File: uast-generator-go/goparser.go [hunk 9]
```diff
@@ -236,18 +249,65 @@ func (t *UastMapper) computeBasicLitKinds(tok token.Token) []Kind {
...
+		// >
+	case token.GTR:
+		return []Kind{OPERATOR, OPERATOR_GREATER_THEN}
```
Comment (ghost) [inline uast-generator-go/goparser.go]: `THEN` -> `THAN`

**Review by ghost — CHANGES_REQUESTED** [review]

File: uast-generator-java/src/main/java/org/sonar/uast/generator/java/UastNode.java (hunk 5 above)
Comment (saberduck) [reply uast-generator-java/src/main/java/org/sonar/uast/generator/java/UastNode.java]: yes, I also extended the approach to cover parenthesis and `if`

**Comment by saberduck** [comment]
@m-g-sonar applied feedback, pls have a look again

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
Nice work!

**Pull request merged**